modules/fee-market/src/s2s/payment.rs

```diff
@@ -290,10 +278,66 @@ pub(crate) fn do_reward<T: Config<I>, I: 'static>(
 	}
 }
 
+/// Record the concrete reward distribution of certain order
+#[derive(Clone, Debug, Encode, Decode, Eq, PartialEq, TypeInfo)]
+pub struct RewardItem<AccountId, Balance> {
+	pub to_slot_relayer: Option<(AccountId, Balance)>,
+	pub to_treasury: Option<Balance>,
+	pub to_message_relayer: Option<(AccountId, Balance)>,
+	pub to_confirm_relayer: Option<(AccountId, Balance)>,
+}
+
+impl<AccountId, Balance> RewardItem<AccountId, Balance> {
+	fn new() -> Self {
+		Self {
+			to_slot_relayer: None,
+			to_treasury: None,
+			to_message_relayer: None,
+			to_confirm_relayer: None,
+		}
+	}
+}
+
 /// Record the calculation rewards result
-pub struct RewardsBook<AccountId, Balance> {
-	pub messages_relayers_rewards: BTreeMap<AccountId, Balance>,
-	pub confirmation_relayer_rewards: Balance,
-	pub assigned_relayers_rewards: BTreeMap<AccountId, Balance>,
-	pub treasury_total_rewards: Balance,
+#[derive(Clone, Debug, Eq, PartialEq, TypeInfo)]
+pub struct RewardsBook<T: Config<I>, I: 'static> {
+	pub deliver_sum: BTreeMap<T::AccountId, BalanceOf<T, I>>,
+	pub confirm_sum: BalanceOf<T, I>,
+	pub assigned_relayers_sum: BTreeMap<T::AccountId, BalanceOf<T, I>>,
+	pub treasury_sum: BalanceOf<T, I>,
+}
+
+impl<T: Config<I>, I: 'static> RewardsBook<T, I> {
+	fn new() -> Self {
+		Self {
+			deliver_sum: BTreeMap::new(),
+			confirm_sum: BalanceOf::<T, I>::zero(),
+			assigned_relayers_sum: BTreeMap::new(),
+			treasury_sum: BalanceOf::<T, I>::zero(),
+		}
+	}
+
+	fn add_reward_item(&mut self, item: RewardItem<T::AccountId, BalanceOf<T, I>>) {
+		if let Some((id, reward)) = item.to_slot_relayer {
+			self.assigned_relayers_sum
+				.entry(id)
+				.and_modify(|r| *r = r.saturating_add(reward))
+				.or_insert(reward);
+		}
+
+		if let Some(reward) = item.to_treasury {
+			self.treasury_sum = self.treasury_sum.saturating_add(reward);
+		}
+
+		if let Some((id, reward)) = item.to_message_relayer {
+			self.deliver_sum
+				.entry(id)
+				.and_modify(|r| *r = r.saturating_add(reward))
+				.or_insert(reward);
+		}
+
+		if let Some((_id, reward)) = item.to_confirm_relayer {
+			self.confirm_sum = self.confirm_sum.saturating_add(reward);
+		}
+	}
```

Comment:
Another way would be directly iterator though RewardItem and calculate the value wanted. RewardBook can be removed then.

Reply:
We can do this. Store the reward item in something like HashMap and iterate it in the `pay_relayers_rewards`. But I think the reward book is easier to read and understand.

Reply:
My feelings from my point as a reviewer, RewardBook is unnecessary concept and removing it will be simpler.

I'm going to merging this and review this discussion in separate issue because its blocking other release tasks.